modules/function/src/main/java/com/opengamma/strata/function/calculation/DefaultLookupRatesProvider.java

```diff
@@ -0,0 +1,280 @@
+/**
+ * Copyright (C) 2016 - present by OpenGamma Inc. and the OpenGamma group of companies
+ *
+ * Please see distribution for license.
+ */
+package com.opengamma.strata.function.calculation;
+
+import java.time.LocalDate;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import org.joda.beans.BeanDefinition;
+import org.joda.beans.ImmutableBean;
+import org.joda.beans.ImmutableConstructor;
+import org.joda.beans.JodaBeanUtils;
+import org.joda.beans.MetaBean;
+import org.joda.beans.Property;
+import org.joda.beans.PropertyDefinition;
+import org.joda.beans.impl.light.LightMetaBean;
+
+import com.opengamma.strata.basics.currency.Currency;
+import com.opengamma.strata.basics.currency.CurrencyPair;
+import com.opengamma.strata.basics.currency.FxRate;
+import com.opengamma.strata.basics.currency.FxRateProvider;
+import com.opengamma.strata.basics.index.FxIndex;
+import com.opengamma.strata.basics.index.IborIndex;
+import com.opengamma.strata.basics.index.Index;
+import com.opengamma.strata.basics.index.OvernightIndex;
+import com.opengamma.strata.basics.index.PriceIndex;
+import com.opengamma.strata.basics.market.MarketData;
+import com.opengamma.strata.basics.market.MarketDataFxRateProvider;
+import com.opengamma.strata.basics.market.MarketDataKey;
+import com.opengamma.strata.collect.ArgChecker;
+import com.opengamma.strata.collect.timeseries.LocalDateDoubleTimeSeries;
+import com.opengamma.strata.market.curve.Curve;
+import com.opengamma.strata.market.curve.CurveName;
+import com.opengamma.strata.market.curve.InterpolatedNodalCurve;
+import com.opengamma.strata.market.id.SimpleCurveId;
+import com.opengamma.strata.market.key.IndexRateKey;
+import com.opengamma.strata.market.view.DiscountFactors;
+import com.opengamma.strata.market.view.DiscountFxForwardRates;
+import com.opengamma.strata.market.view.DiscountFxIndexRates;
+import com.opengamma.strata.market.view.FxForwardRates;
+import com.opengamma.strata.market.view.FxIndexRates;
+import com.opengamma.strata.market.view.IborIndexRates;
+import com.opengamma.strata.market.view.OvernightIndexRates;
+import com.opengamma.strata.market.view.PriceIndexValues;
+import com.opengamma.strata.pricer.rate.AbstractRatesProvider;
+import com.opengamma.strata.pricer.rate.RatesProvider;
+
+/**
+ * A rates provider based on a rates lookup.
+ * <p>
+ * This uses a {@link DefaultRatesMarketDataLookup} to provide a view on {@link MarketData}.
+ */
+@BeanDefinition(style = "light")
+final class DefaultLookupRatesProvider
+    extends AbstractRatesProvider
+    implements RatesProvider, ImmutableBean {
+
+  /**
+   * The lookup.
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final DefaultRatesMarketDataLookup lookup;
```

Comment:
The maps of discount currency to identifier and forward index to identifier are not public on the interface (as the interface could be implemented in other ways. This class is essentially an inner class of `DefaultRatesMarketDataLookup` and is package scope, so it is fine as is.